# It is a regression problem 

## We are doing the machine learning on house price dataset

<table align="center" width=100%>
    <tr>
        </td>
            </td>
        </td>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                    <b>Regression modelling on House Price Data
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

# Problem Statement

This dataset is created for the prediction on House Prices from an Indian perspective towns. The dataset contains various features that are important during the application for the selling the Houses. The predicted output obtained from the classification algorithm gives a fair idea about the chances of a Houses Sold. 

In [158]:
# import 'Pandas'
# It is utilised to format the data structure. It means it can convert any data structure into pandas data frame.
import pandas as pd

# import 'numpy'
# It is used to find statistical summary.
import numpy as np

# import subpackages of Matplotlib
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap

# import 'seaborn'
import seaborn as sns

In [160]:
# Load the data and name it df_houseprice
# Load the csv file
# store the data in 'df_houseprice'
df_houseprice = pd.read_csv("C:\\Users\\HP\\Downloads\\House_Price1.csv")

# display first five observation using head()
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

## Data Overview

In [163]:
# Name the Columns 
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

**Interpretation**

 - `price`:Price of the House.
 - `resid_area`:Residential Area where my house is located.
 - `air_qual`:Air quality of the atmosphere.
 - `room_num`:The plot no.of the House.
 - `age`:The age of the people buying the house.
 - `distance`:The distance of the school

In [166]:
# Basic Info of the Data Set
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        506 non-null    float64
 1   resid_area   506 non-null    float64
 2   air_qual     506 non-null    float64
 3   room_num     506 non-null    float64
 4   age          506 non-null    float64
 5   distance     506 non-null    float64
 6   teachers     506 non-null    float64
 7   poor_prop    506 non-null    float64
 8   airport      506 non-null    object 
 9   n_hos_beds   498 non-null    float64
 10  n_hot_rooms  506 non-null    float64
 11  waterbody    351 non-null    object 
 12  rainfall     506 non-null    int64  
 13  bus_ter      506 non-null    object 
 14  parks        506 non-null    float64
 15  Sold         506 non-null    int64  
dtypes: float64(11), int64(2), object(3)
memory usage: 63.4+ KB


**Interpretation**

 - we have 13 numerical column and 3 categorial column
 - we have null values in the data
 - we have total 16 columns in the data

In [169]:
# Shape of the Data
df_houseprice.shape

(506, 16)

**Interpretation**
 - We have total 16 columns.
 - We have 506 observations.

In [172]:
# Summarize the basic statistics of Data
df_houseprice.describe().T

count       mean        std        min        25%        50%  \
price        506.0  22.528854   9.182176   5.000000  17.025000  21.200000   
resid_area   506.0  41.136779   6.860353  30.460000  35.190000  39.690000   
air_qual     506.0   0.554695   0.115878   0.385000   0.449000   0.538000   
room_num     506.0   6.284634   0.702617   3.561000   5.885500   6.208500   
age          506.0  68.574901  28.148861   2.900000  45.025000  77.500000   
distance     506.0   3.971996   2.108532   1.130000   2.270000   3.385000   
teachers     506.0  21.544466   2.164946  18.000000  19.800000  20.950000   
poor_prop    506.0  12.653063   7.141062   1.730000   6.950000  11.360000   
n_hos_beds   498.0   7.899767   1.476683   5.268000   6.634500   7.999000   
n_hot_rooms  506.0  13.041605   5.238957  10.057600  11.189800  12.720000   
rainfall     506.0  39.181818  12.513697   3.000000  28.000000  39.000000   
parks        506.0   0.054454   0.010632   0.033292   0.046464   0.053507   
Sold         506.0   0.454545   0.498422   0.000000   0.000000   0.000000   

                   75%         max  
price        25.000000   50.000000  
resid_area   48.100000   57.740000  
air_qual      0.624000    0.871000  
room_num      6.623500    8.780000  
age          94.075000  100.000000  
distance      5.367500   12.320000  
teachers     22.600000   27.400000  
poor_prop    16.955000   37.970000  
n_hos_beds    9.088000   10.876000  
n_hot_rooms  14.170800  101.120000  
rainfall     50.000000   60.000000  
parks         0.061397    0.086711  
Sold          1.000000    1.000000

**Interpretation**
 - We have outliers in the data, most of the outliers are in age column.

## Data Preprocessing

### a. Data Cleaning

In [177]:
# Show the first five observations
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

In [179]:
# Rename the Columns
# Step 1 : Show the column names 
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

In [181]:
# Step 2 : Allocate new column names to old column names
# Make the dictionary{'Old column name':'New column name'}
d1={'price':'Price',
    'resid_area':'Residential_Area',
    'air_qual':'Air_Quality',
    'room_num':'Room_Numbers',
    'age':'Age',
    'distance':'Distance',
    'teachers':'Teachers',
    'poor_prop':'Proportion_Of_Poor_Population',
    'airport':'Airport',
    'n_hos_beds':'Number_Of_Hospital_Beds',
    'n_hot_rooms':'Number_Of_Hotel_Rooms',
    'waterbody':'Waterbody',
    'rainfall':'Rainfall',
    'bus_ter':'Bus_Terminal',
    'parks':'Parks',
    'Sold':'Sold'}

In [183]:
df_houseprice = df_houseprice.rename(columns=d1)

In [185]:
# Authenticate the Results
df_houseprice.columns

Index(['Price', 'Residential_Area', 'Air_Quality', 'Room_Numbers', 'Age',
       'Distance', 'Teachers', 'Proportion_Of_Poor_Population', 'Airport',
       'Number_Of_Hospital_Beds', 'Number_Of_Hotel_Rooms', 'Waterbody',
       'Rainfall', 'Bus_Terminal', 'Parks', 'Sold'],
      dtype='object')

### b. Null Value Handling

In [188]:
# Find the valid Null Values
df_houseprice.isnull().sum()

Price                              0
Residential_Area                   0
Air_Quality                        0
Room_Numbers                       0
Age                                0
Distance                           0
Teachers                           0
Proportion_Of_Poor_Population      0
Airport                            0
Number_Of_Hospital_Beds            8
Number_Of_Hotel_Rooms              0
Waterbody                        155
Rainfall                           0
Bus_Terminal                       0
Parks                              0
Sold                               0
dtype: int64

In [190]:
# Percentage missing value
round((df_houseprice.isnull().sum())/len(df_houseprice)*100,3)

Price                             0.000
Residential_Area                  0.000
Air_Quality                       0.000
Room_Numbers                      0.000
Age                               0.000
Distance                          0.000
Teachers                          0.000
Proportion_Of_Poor_Population     0.000
Airport                           0.000
Number_Of_Hospital_Beds           1.581
Number_Of_Hotel_Rooms             0.000
Waterbody                        30.632
Rainfall                          0.000
Bus_Terminal                      0.000
Parks                             0.000
Sold                              0.000
dtype: float64

In [192]:
# Find the nature of the column
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Price                          506 non-null    float64
 1   Residential_Area               506 non-null    float64
 2   Air_Quality                    506 non-null    float64
 3   Room_Numbers                   506 non-null    float64
 4   Age                            506 non-null    float64
 5   Distance                       506 non-null    float64
 6   Teachers                       506 non-null    float64
 7   Proportion_Of_Poor_Population  506 non-null    float64
 8   Airport                        506 non-null    object 
 9   Number_Of_Hospital_Beds        498 non-null    float64
 10  Number_Of_Hotel_Rooms          506 non-null    float64
 11  Waterbody                      351 non-null    object 
 12  Rainfall                       506 non-null    int

In [194]:
# This is a categorical column so we can replace the null with mode.
df_houseprice.Waterbody.mode()

0    River
Name: Waterbody, dtype: object

In [196]:
# Replace the null in that column
df_houseprice.Waterbody = df_houseprice.Waterbody.fillna('River')

In [198]:
# Percentage missing value
round((df_houseprice.isnull().sum())/len(df_houseprice)*100,3)

Price                            0.000
Residential_Area                 0.000
Air_Quality                      0.000
Room_Numbers                     0.000
Age                              0.000
Distance                         0.000
Teachers                         0.000
Proportion_Of_Poor_Population    0.000
Airport                          0.000
Number_Of_Hospital_Beds          1.581
Number_Of_Hotel_Rooms            0.000
Waterbody                        0.000
Rainfall                         0.000
Bus_Terminal                     0.000
Parks                            0.000
Sold                             0.000
dtype: float64

In [200]:
# In Number_Of_Hospital_Beds contain less than 10% values so we can drop null values
df_houseprice.dropna(inplace=True)

In [202]:
# Percentage missing value
round((df_houseprice.isnull().sum())/len(df_houseprice)*100,3)

Price                            0.0
Residential_Area                 0.0
Air_Quality                      0.0
Room_Numbers                     0.0
Age                              0.0
Distance                         0.0
Teachers                         0.0
Proportion_Of_Poor_Population    0.0
Airport                          0.0
Number_Of_Hospital_Beds          0.0
Number_Of_Hotel_Rooms            0.0
Waterbody                        0.0
Rainfall                         0.0
Bus_Terminal                     0.0
Parks                            0.0
Sold                             0.0
dtype: float64

## Prepare Data for Machine Learning

In [205]:
# We know that the Machine Learning Alorithm are Mathemaitical Expression so we need to provide the data to the algorithm in specific format
# Encoding

In [207]:
# Step 1 : Divide the detabase on their datatype
df_character = df_houseprice.select_dtypes(include = ['object'])
df_character.columns

Index(['Airport', 'Waterbody', 'Bus_Terminal'], dtype='object')

In [233]:
# Stpe 2 : Use dummy encoding to convert character columns into integer column
dummy_var = pd.get_dummies(data = df_character, drop_first = True)
dummy_var = dummy_var.reset_index()
dummy_var2 = pd.get_dummies(df_character)

In [235]:
dummy_var.head()

index  Airport_YES  Waterbody_Lake and River  Waterbody_River
0      0         True                     False             True
1      1        False                     False            False
2      2        False                     False             True
3      3         True                     False            False
4      4        False                     False            False

In [237]:
dummy_var2.head()

Airport_NO  Airport_YES  Waterbody_Lake  Waterbody_Lake and River  \
0       False         True           False                     False   
1        True        False            True                     False   
2        True        False           False                     False   
3       False         True            True                     False   
4        True        False            True                     False   

   Waterbody_River  Bus_Terminal_YES  
0             True              True  
1            False              True  
2             True              True  
3            False              True  
4            False              True

In [215]:
df_number = df_houseprice.select_dtypes(include = [np.number])
df_number.head()

Price  Residential_Area  Air_Quality  Room_Numbers   Age  Distance  \
0   24.0             32.31        0.538         6.575  65.2      4.35   
1   21.6             37.07        0.469         6.421  78.9      4.99   
2   34.7             37.07        0.469         7.185  61.1      5.03   
3   33.4             32.18        0.458         6.998  45.8      6.21   
4   36.2             32.18        0.458         7.147  54.2      6.16   

   Teachers  Proportion_Of_Poor_Population  Number_Of_Hospital_Beds  \
0      24.7                           4.98                    5.480   
1      22.2                           9.14                    7.332   
2      22.2                           4.03                    7.394   
3      21.3                           2.94                    9.268   
4      21.3                           5.33                    8.824   

   Number_Of_Hotel_Rooms  Rainfall     Parks  Sold  
0                11.1920        23  0.049347     0  
1                12.1728        42  0.046146     1  
2               101.1200        38  0.045764     0  
3                11.2672        45  0.047151     0  
4                11.2896        55  0.039474     0

In [217]:
from sklearn.preprocessing import StandardScaler
df_target = df_houseprice['Price']
df_number = df_number.drop('Price',axis = 1)
df_number.head()

Residential_Area  Air_Quality  Room_Numbers   Age  Distance  Teachers  \
0             32.31        0.538         6.575  65.2      4.35      24.7   
1             37.07        0.469         6.421  78.9      4.99      22.2   
2             37.07        0.469         7.185  61.1      5.03      22.2   
3             32.18        0.458         6.998  45.8      6.21      21.3   
4             32.18        0.458         7.147  54.2      6.16      21.3   

   Proportion_Of_Poor_Population  Number_Of_Hospital_Beds  \
0                           4.98                    5.480   
1                           9.14                    7.332   
2                           4.03                    7.394   
3                           2.94                    9.268   
4                           5.33                    8.824   

   Number_Of_Hotel_Rooms  Rainfall     Parks  Sold  
0                11.1920        23  0.049347     0  
1                12.1728        42  0.046146     1  
2               101.1200        38  0.045764     0  
3                11.2672        45  0.047151     0  
4                11.2896        55  0.039474     0

In [219]:
# Initiate the Standard Scaler
ss = StandardScaler()

In [221]:
# Step 3 : Transform the data 
Scale = ss.fit_transform(df_number)
Scale

array([[-1.28244435, -0.13709359,  0.40897478, ..., -1.30206786,
        -0.47313205, -0.91152748],
       [-0.59005882, -0.73319492,  0.18997385, ...,  0.21519397,
        -0.77522463,  1.09705963],
       [-0.59005882, -0.73319492,  1.276446  , ..., -0.10422957,
        -0.81123666, -0.91152748],
       ...,
       [ 0.11687263,  0.16527665,  0.97923045, ..., -0.66322077,
         0.30293265,  1.09705963],
       [ 0.11687263,  0.16527665,  0.72041117, ...,  0.6144734 ,
         0.59750894,  1.09705963],
       [ 0.11687263,  0.16527665, -0.36606097, ...,  0.45476163,
         0.56369225,  1.09705963]])

In [239]:
df_scale = pd.DataFrame(Scale, columns = df_number.columns)
df_scale = df_scale.reset_index()
df_scale.head()

index  Residential_Area  Air_Quality  Room_Numbers       Age  Distance  \
0      0         -1.282444    -0.137094      0.408975 -0.115644  0.172845   
1      1         -0.590059    -0.733195      0.189974  0.369741  0.476100   
2      2         -0.590059    -0.733195      1.276446 -0.260905  0.495053   
3      3         -1.301354    -0.828226      1.010516 -0.802977  1.054179   
4      4         -1.301354    -0.828226      1.222407 -0.505369  1.030487   

   Teachers  Proportion_Of_Poor_Population  Number_Of_Hospital_Beds  \
0  1.462167                      -1.066368                -1.640299   
1  0.304963                      -0.484136                -0.384875   
2  0.304963                      -1.199330                -0.342847   
3 -0.111631                      -1.351886                 0.927490   
4 -0.111631                      -1.017382                 0.626514   

   Number_Of_Hotel_Rooms  Rainfall     Parks      Sold  
0              -0.352111 -1.302068 -0.473132 -0.911527  
1              -0.166019  0.215194 -0.775225  1.097060  
2              16.710340 -0.104230 -0.811237 -0.911527  
3              -0.337843  0.454762 -0.680402 -0.911527  
4              -0.333593  1.253320 -1.404723 -0.911527

In [241]:
# Step 4 : Make a DataFrame for independent variables
df_independent = pd.concat([df_scale, dummy_var],axis = 1)
df_independent.head()

index  Residential_Area  Air_Quality  Room_Numbers       Age  Distance  \
0      0         -1.282444    -0.137094      0.408975 -0.115644  0.172845   
1      1         -0.590059    -0.733195      0.189974  0.369741  0.476100   
2      2         -0.590059    -0.733195      1.276446 -0.260905  0.495053   
3      3         -1.301354    -0.828226      1.010516 -0.802977  1.054179   
4      4         -1.301354    -0.828226      1.222407 -0.505369  1.030487   

   Teachers  Proportion_Of_Poor_Population  Number_Of_Hospital_Beds  \
0  1.462167                      -1.066368                -1.640299   
1  0.304963                      -0.484136                -0.384875   
2  0.304963                      -1.199330                -0.342847   
3 -0.111631                      -1.351886                 0.927490   
4 -0.111631                      -1.017382                 0.626514   

   Number_Of_Hotel_Rooms  Rainfall     Parks      Sold  index  Airport_YES  \
0              -0.352111 -1.302068 -0.473132 -0.911527      0         True   
1              -0.166019  0.215194 -0.775225  1.097060      1        False   
2              16.710340 -0.104230 -0.811237 -0.911527      2        False   
3              -0.337843  0.454762 -0.680402 -0.911527      3         True   
4              -0.333593  1.253320 -1.404723 -0.911527      4        False   

   Waterbody_Lake and River  Waterbody_River  
0                     False             True  
1                     False            False  
2                     False             True  
3                     False            False  
4                     False            False

In [243]:
# Step 5 : Split the data for training  and testing
import statsmodels
import statsmodels.api as sm

df_independent = sm.add_constant(df_independent)
df_independent.head()

const  index  Residential_Area  Air_Quality  Room_Numbers       Age  \
0    1.0      0         -1.282444    -0.137094      0.408975 -0.115644   
1    1.0      1         -0.590059    -0.733195      0.189974  0.369741   
2    1.0      2         -0.590059    -0.733195      1.276446 -0.260905   
3    1.0      3         -1.301354    -0.828226      1.010516 -0.802977   
4    1.0      4         -1.301354    -0.828226      1.222407 -0.505369   

   Distance  Teachers  Proportion_Of_Poor_Population  Number_Of_Hospital_Beds  \
0  0.172845  1.462167                      -1.066368                -1.640299   
1  0.476100  0.304963                      -0.484136                -0.384875   
2  0.495053  0.304963                      -1.199330                -0.342847   
3  1.054179 -0.111631                      -1.351886                 0.927490   
4  1.030487 -0.111631                      -1.017382                 0.626514   

   Number_Of_Hotel_Rooms  Rainfall     Parks      Sold  index  Airport_YES  \
0              -0.352111 -1.302068 -0.473132 -0.911527      0         True   
1              -0.166019  0.215194 -0.775225  1.097060      1        False   
2              16.710340 -0.104230 -0.811237 -0.911527      2        False   
3              -0.337843  0.454762 -0.680402 -0.911527      3         True   
4              -0.333593  1.253320 -1.404723 -0.911527      4        False   

   Waterbody_Lake and River  Waterbody_River  
0                     False             True  
1                     False            False  
2                     False             True  
3                     False            False  
4                     False            False

## Testing

In [246]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(df_independent, df_target, random_state = 10, test_size = 0.2)

In [248]:
# print dimensios of train set
print('X_train', X_train.shape)
print('y_train', y_train.shape)

# print dimensions of test set
print('X_test', X_test.shape)
print('y_test', y_test.shape)

X_train (398, 18)
y_train (398,)
X_test (100, 18)
y_test (100,)


In [250]:
X_train.columns

Index(['const', 'index', 'Residential_Area', 'Air_Quality', 'Room_Numbers',
       'Age', 'Distance', 'Teachers', 'Proportion_Of_Poor_Population',
       'Number_Of_Hospital_Beds', 'Number_Of_Hotel_Rooms', 'Rainfall', 'Parks',
       'Sold', 'index', 'Airport_YES', 'Waterbody_Lake and River',
       'Waterbody_River'],
      dtype='object')

### 1. Linear Regression

In [254]:
from sklearn.linear_model import LinearRegression

# Create a linear regression model
model = LinearRegression()

#Fit the model to the training data
model.fit(X_train, y_train)

# Predict on the testing data
predictions = model.predict(X_test)

In [256]:
from sklearn.metrics import mean_squared_error

# Calculate mean squared error
mse = mean_squared_error(y_test,predictions)
print("Mean Squared Error : " , mse)

Mean Squared Error :  15.498182384260335


In [258]:
from sklearn.metrics import mean_absolute_error


# Calculate mean absolute error
mae = mean_absolute_error(y_test,predictions)
print("Mean Absolute Error : " , mae)

Mean Absolute Error :  3.133142913286591


In [260]:
from sklearn.metrics import root_mean_squared_error


# Calculate Root Mean Squared error
rmse = root_mean_squared_error(y_test,predictions)
print("Root Mean Squared Error : " , rmse)

Root Mean Squared Error :  3.9367730928084153
